### PR TITLE
[WIP] Rootless docker

### PR DESCRIPTION
**Pull request by sapk**
Following #4749 

Currently work with the default user git pid (1000)

The main blocking adjustment to use the [--user docker flag](https://docs.docker.com/engine/reference/run/#user) to adjust running user is various check of the user by gitea but also by sub process like ssh-keygen in #7128

### CODE REVIEW

File: docker/root/usr/bin/setup [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #!/bin/bash
 
+#TODO maybe /data/git/.ssh not needed anymore
```
Comment (lunny) [inline docker/root/usr/bin/setup:3]: If docker user internal SSH server, it's not needed

File: docker/root/usr/bin/setup [hunk 2]
```diff
@@ -41,12 +44,7 @@ if [ ! -f ${GITEA_CUSTOM}/conf/app.ini ]; then
...
-    chown ${USER}:git ${GITEA_CUSTOM}/conf/app.ini
 fi
 
-# only chown if current owner is not already the gitea ${USER}. No recursive check to save time
-if ! [[ $(ls -ld /data/gitea | awk '{print $3}') = ${USER} ]]; then chown -R ${USER}:git /data/gitea; fi
-if ! [[ $(ls -ld /app/gitea  | awk '{print $3}') = ${USER} ]]; then chown -R ${USER}:git /app/gitea;  fi
-if ! [[ $(ls -ld /data/git   | awk '{print $3}') = ${USER} ]]; then chown -R ${USER}:git /data/git;   fi
-chmod 0755 /data/gitea /app/gitea /data/git
+chmod 0755 /data/gitea /data/git
+#chmod 0755 /app/gitea
```
Comment (das7pad) [inline docker/root/usr/bin/setup:50]: `/app/gitea` is owned by root.
```suggestion
```

File: models/ssh_key.go [hunk 3]
```diff
@@ -361,18 +361,11 @@ func checkKeyFingerprint(e Engine, fingerprint string) error {
...
 	}
-	defer os.Remove(tmpPath)
-	stdout, stderr, err := process.GetManager().Exec("AddPublicKey", "ssh-keygen", "-lf", tmpPath)
-	if err != nil {
-		return "", fmt.Errorf("'ssh-keygen -lf %s' failed with error '%s': %s", tmpPath, err, stderr)
-	} else if len(stdout) < 2 {
-		return "", errors.New("not enough output for calculating fingerprint: " + stdout)
-	}
-	return strings.Split(stdout, " ")[1], nil
+	return ssh.FingerprintSHA256(pk), nil
```
Comment (das7pad) [inline models/ssh_key.go]: This seems to be an unrelated change. Can we move the change into a new PR?

File: modules/setting/setting.go [hunk 4]
```diff
@@ -436,7 +436,7 @@ func forcePathSeparator(path string) {
 // This check is ignored under Windows since SSH remote login is not the main
 // method to login on Windows.
 func IsRunUserMatchCurrentUser(runUser string) (string, bool) {
-	if IsWindows {
+	if IsWindows || SSH.StartBuiltinServer {
```
Comment (das7pad) [inline modules/setting/setting.go]: Seems to be a bug fix? Let's move it into a new PR.

File: docker/root/usr/bin/setup (hunk 1 above)
Comment (das7pad) [reply docker/root/usr/bin/setup:3]: While it is not needed for the single process setup, removing it would break existing setups.

**Review by das7pad — CHANGES_REQUESTED** [review]
This breaks almost every usage of the docker image.

Some broken usage examples:
 - all of the 'official' docker-compose configs
  https://docs.gitea.io/en-us/install-with-docker/
 - `docker run -d -p 80:3000 -p 2222:22 gitea/gitea`
 - running sshd in the container to get the gitea binary
   - run gitea with git user
   - run sshd with root

**Comment by das7pad** [comment]
How about adding a new image variant for the proposed changes?

For example `gitea/gitea:root-less`, `gitea/gitea:1-root-less`, `gitea/gitea:1.8-root-less`

This would allow any existing deployment to stay as is. And folks which require additional hardening of the container can use the new variant.

With a reasonable delay we can eventually swap the default tag: `gitea/gitea:root-less`-> `gitea/gitea:latest`.

File: docker/root/usr/bin/setup (hunk 1 above)
Comment (sapk) [reply docker/root/usr/bin/setup:3]: @das7pad yes but it is for the setup part so if we don't need it anymore we don't need to create it if it doesn't exist.

File: docker/root/usr/bin/setup (hunk 2 above)
Comment (sapk) [reply docker/root/usr/bin/setup:50]: Yes and I don't think it is a problem. I will recheck but that a good point that the process inside the containre couldn't change content inside /app/gitea

File: models/ssh_key.go (hunk 3 above)
Comment (sapk) [reply models/ssh_key.go]: It is a needed change because openssh does un-needed system check (that fail in this configuration with --user flag) for calculating the fingerprint of the key. I allready separate this in an other PR #7128

File: modules/setting/setting.go (hunk 4 above)
Comment (sapk) [reply modules/setting/setting.go]: Done in #7215

**Comment by sapk** [comment]
@das7pad the docs will need to be updated to follow some changed needed to the run command in some cases but it should be able to reuse the same data folder. 
I like the idea to have a transition tag to not enforce this rootless image at start.

File: docker/root/usr/bin/setup (hunk 1 above)
Comment (das7pad) [reply docker/root/usr/bin/setup:3]: For folks who backup the entire /data/git directory it is not a problem. But for those who just backup non-generated files, read only the repositories directory, the .ssh directory would be missing upon restore.

File: models/ssh_key.go (hunk 3 above)
Comment (das7pad) [reply models/ssh_key.go]: What system check is failing for you?

I setup a test with alpine and the `--user` flag, but can not see any error.

https://gist.github.com/das7pad/62b58bfb67da516542992d8bf5744446

File: docker/root/usr/bin/setup (hunk 1 above)
Comment (zeripath) [reply docker/root/usr/bin/setup:3]: But the internal sshserver doesn't use the .ssh directory...

Comment (das7pad) [reply docker/root/usr/bin/setup:3]: I was thinking about the third example in https://github.com/go-gitea/gitea/pull/7129#pullrequestreview-250188735.
I know that the files in the .ssh directory are only needed to get the spawned sshd-childs talk with the correct gitea instance.

File: docker/root/etc/s6/openssh/setup [hunk 5]
```diff
@@ -1,29 +0,0 @@
...
-fi
-
-if [ ! -f /data/ssh/ssh_host_ecdsa_key ]; then
-    echo "Generating /data/ssh/ssh_host_ecdsa_key..."
-    ssh-keygen -t ecdsa -b 256 -f /data/ssh/ssh_host_ecdsa_key -N "" > /dev/null
-fi
-
-chown root:root /data/ssh/*
-chmod 0700 /data/ssh
-chmod 0600 /data/ssh/*
```
Comment (das7pad) [inline docker/root/etc/s6/openssh/setup:29]: The ssh-hostkeys are owned by root and locked down to root only read access. The internal ssh-server will not be able to read them as git user.

**Review by das7pad — CHANGES_REQUESTED** [review]
Manual ownership changes to the ssh directory in the data volume are needed.

File: models/ssh_key.go (hunk 3 above)
Comment (sapk) [reply models/ssh_key.go]: If you pass an uid/gid (numerical) that doesn't exist inside the container it should failed.

**Comment by westurner** [comment]
Rather than try and maintain/rebase this separate branch, would it be easier to:
- leave the existing config in
- check for an environment variable like `GITEA_ROOTLESS` in the setup and entrypoint scripts
- create a `templates/app_rootless.ini`

**Comment by sapk** [comment]
@westurner I would suggest more to define a rootless tagged image that could become the main later but would ease the transition and allow us to break some pattern existing in the current image.

**Comment by alexanderadam** [comment]
> @westurner I would suggest more to define a rootless tagged image that could become the main later but would ease the transition and allow us to break some pattern existing in the current image.

I like this idea.

When the rootless image is working (i.e. two months without new issues on the rootless image) the 'legacy root' image could show a warning in the frontend (i.e. `Hey, we are now switching to a proper rootless image. You should migrate, too. Click here for more information on how to migrate`).

**Comment by sapk** [comment]
Yes and this transition duration would help us write a migration guide (like changing owner, possible change of key signature of ssh server, ...)

**Comment by sapk** [comment]
I will send another PR that will replace this one for a image tagged rootless

**Comment by sapk** [comment]
The fixes related to help remove external deps (openssh) are already merged (in separe PR) and the rootless part of this PR is moved to a new PR (non breaking by using an other tag -rootless)
